osf/management/commands/force_archive.py

```diff
@@ -0,0 +1,344 @@
+            manually_archive(filenode['children'], reg, parent=cloned)
+
+
+def modify_file_tree_recursive(tree, file_obj, deleted, cached=False):
+    # Note: `deleted` has three possible states:
+    # - True: indicates file should be marked as deleted
+    # - False: indicates file should be added or undeleted
+    # - None: indicates file version should be reverted by one
+    retree = []
+    noop = True
+    if not isinstance(tree, list):
+        tree = [tree]
+    for filenode in tree:
+        if file_obj.is_deleted and not cached and filenode['object'].id == file_obj.parent.id:
+            filenode['children'].append({
+                'deleted': None,
+                'object': file_obj,
+                'version': int(file_obj.versions.last().identifier) if file_obj.versions.exists() else None
+            })
+            cached = True
+        if filenode['object']._id == file_obj._id:
+            if deleted is not None:
+                filenode['deleted'] = deleted
+                noop = False
+            elif deleted is None:
+                if not isinstance(filenode['version'], int):
+                    raise Exception('Unexpected type for version: got {}'.format(type(filenode['version'])))
+                filenode['version'] = filenode['version'] - 1
+                noop = False
+        if filenode.get('children'):
+            filenode['children'], noop = modify_file_tree_recursive(filenode['children'], file_obj, deleted, cached)
+        retree.append(filenode)
+    assert not noop
+    return retree, noop
+
+def revert_log_actions(file_tree, reg, obj_cache):
+    logs_to_revert = reg.registered_from.logs.filter(date__gt=reg.registered_date).exclude(action__in=LOG_WHITELIST).order_by('-date')
+    for log in list(logs_to_revert):
+        file_obj = reg.registered_from.files.get(_id=log.params['urls']['view'].split('/')[5])
+        if log.action == 'osf_storage_file_added':
+            # Find and mark deleted
+            logger.info('Reverting add {}:{} from {}'.format(file_obj._id, file_obj.name, log.date))
+            file_tree, _ = modify_file_tree_recursive(file_tree, file_obj, deleted=True, cached=bool(file_obj._id in obj_cache))
+        elif log.action == 'osf_storage_file_removed':
+            # Find parent and add to children, or undelete
+            logger.info('Reverting delete {}:{} from {}'.format(file_obj._id, file_obj.name, log.date))
+            file_tree, _ = modify_file_tree_recursive(file_tree, file_obj, deleted=False, cached=bool(file_obj._id in obj_cache))
+        elif log.action == 'osf_storage_file_updated':
+            # Find file and revert version
+            logger.info('Reverting update {}:{} from {}'.format(file_obj._id, file_obj.name, log.date))
+            file_tree, _ = modify_file_tree_recursive(file_tree, file_obj, deleted=None, cached=bool(file_obj._id in obj_cache))
+            pass
+        elif log.action == 'osf_storage_folder_created':
```

Comment:
Could, but I intentionally kept these separate for logging. I can change that if you don't think it matters